promql/engine_test.go

```diff
@@ -1634,6 +1634,217 @@ load 1ms
+			if c.interval == 0 {
+				c.interval = 1
+			}
+			start, end, interval := time.Unix(c.start, 0), time.Unix(c.end, 0), time.Duration(c.interval)*time.Second
+			var err error
+			var qry Query
+			if c.end == 0 {
+				qry, err = engine.NewInstantQuery(context.Background(), storage, nil, c.query, start)
+			} else {
+				qry, err = engine.NewRangeQuery(context.Background(), storage, nil, c.query, start, end, interval)
+			}
+			require.NoError(t, err)
+
+			res := qry.Exec(context.Background())
+			require.NoError(t, res.Err)
+			switch {
+			case c.result != nil:
+				if expMat, ok := c.result.(Matrix); ok {
+					sort.Sort(expMat)
+					sort.Sort(res.Value.(Matrix))
+				}
+				require.Equal(t, c.result, res.Value, "query %q failed for require.Equal()", c.query)
+			case c.resultIn != nil:
+				require.Equal(t, c.resultLen, len(res.Value.(Matrix)), "query %q failed", c.query)
+				if expMat, ok := c.resultIn.(Matrix); ok {
+					sort.Sort(expMat)
+					sort.Sort(res.Value.(Matrix))
+					for _, e := range res.Value.(Matrix) {
+						require.Contains(t, c.resultIn, e, "query %q failed for require.Contains()", c.query)
+					}
+				}
+			}
+		})
+	}
+
+	// Dynamically build sample_ratio() cases, to verify that
+	//   sample_ratio(v, metric)
+	// is the *exact complement* of
+	//   sample_ratio(-(1.0-v), metric)
+	//
+	// e.g. v=0.2 is the complement (set of timeseries) of v=-0.8
+	type ratioCase struct {
+		query                string
+		start, end, interval int64 // Time in seconds.
+		result               parser.Value
+	}
+	orQuery := `sample_ratio(%f, metric) or sample_ratio(%f, metric)`
+	andQuery := `sample_ratio(%f, metric) and sample_ratio(%f, metric)`
+	var ratioCases []ratioCase
+	for v := 0.0; v < 1.0; v += 0.1 {
+		// OR-ing both must return the full matrix
+		ratioCases = append(ratioCases, ratioCase{
+			query: fmt.Sprintf(orQuery, v, -(1.0 - v)),
+			start: 0, end: 20, interval: 10,
+			result: fullMatrix20,
+		})
+		// AND-ing both must return an empty matrix
+		ratioCases = append(ratioCases, ratioCase{
+			query: fmt.Sprintf(andQuery, v, -(1.0 - v)),
+			start: 0, end: 20, interval: 10,
+			result: Matrix{},
+		})
+	}
+
+	for _, c := range ratioCases {
+		t.Run(c.query, func(t *testing.T) {
+			if c.interval == 0 {
+				c.interval = 1
+			}
+			start, end, interval := time.Unix(c.start, 0), time.Unix(c.end, 0), time.Duration(c.interval)*time.Second
+			var err error
+			var qry Query
+			if c.end == 0 {
+				qry, err = engine.NewInstantQuery(context.Background(), storage, nil, c.query, start)
+			} else {
+				qry, err = engine.NewRangeQuery(context.Background(), storage, nil, c.query, start, end, interval)
+			}
+			require.NoError(t, err)
+
+			res := qry.Exec(context.Background())
+			require.NoError(t, res.Err)
+			if expMat, ok := c.result.(Matrix); ok {
+				sort.Sort(expMat)
+				sort.Sort(res.Value.(Matrix))
+			}
+			require.Equal(t, c.result, res.Value, "query %q failed", c.query)
+		})
+	}
+}
```

Comment:
**Suggestion:** Consider refactoring the test cases for the `sample_limit` and `sample_ratio` functions into separate test functions. This will improve the readability and maintainability of the tests.
```suggestion
func TestSampleLimit(t *testing.T) {
	engine := newTestEngine()
	storage := LoadedStorage(t, `
load 10s
  metric{job="1"} 0+1x1000
  metric{job="2"} 0+2x1000
  metric{job="3"} 0+3x1000
  metric{job="4"} 0+4x1000
  metric{job="5"} 0+5x1000
`)
	t.Cleanup(func() { storage.Close() })
	...
	cases := []struct {
		query                string
		start, end, interval int64 // Time in seconds.
		result               parser.Value
		resultIn             parser.Value
		resultLen            int // Required if resultIn is set
	}{
		...
	}
	...
	for _, c := range cases {
		t.Run(c.query, func(t *testing.T) {
			...
		})
	}
}

func TestSampleRatio(t *testing.T) {
	engine := newTestEngine()
	storage := LoadedStorage(t, `
load 10s
  metric{job="1"} 0+1x1000
  metric{job="2"} 0+2x1000
  metric{job="3"} 0+3x1000
  metric{job="4"} 0+4x1000
  metric{job="5"} 0+5x1000
`)
	t.Cleanup(func() { storage.Close() })
	...
	cases := []struct {
		query                string
		start, end, interval int64 // Time in seconds.
		result               parser.Value
		resultIn             parser.Value
		resultLen            int // Required if resultIn is set
	}{
		...
	}
	...
	for _, c := range cases {
		t.Run(c.query, func(t *testing.T) {
			...
		})
	}
}
```